### ISSUE TITLE
False error for exhaustivity check over tuples

### ISSUE THREAD

**Issue by kyle-query** [microsoft/pyright#6591]
**Describe the bug**

pyright does not correctly analyze exhaustivity of `match` statements over tuples:

```python
from enum import Enum

class X(Enum):
    A = 1
    B = 2

def foo(x: X, y: X):
    # Cases within match statement do not exhaustively handle all value
    #     Unhandled type: "tuple[X, X]"
    #     If exhaustive handling is not intended, add "case _: pass"
    match (x, y):
        case X.A, X.A:
            pass
        case X.A, X.B:
            pass
        case X.B, X.A:
            pass
        case X.B, X.B:
            pass
```

This also occurs over tuples of Unions.

**VS Code extension or command-line**

Running pyright 1.1.335 on the command line.

**Comment by erictraut** [comment]
This is intended behavior, so I don't consider it a bug.

Refer to [this issue](https://github.com/microsoft/pyright/issues/6437) for details.

**Comment by kyle-query** [comment]
@robert-bn, I had the same thought as I was rewriting my match statements. With only two-element tuples in my case it wasn't a big deal, but more deeply nested match blocks would become hard to read.

I'm just spitballing, but maybe the concern about combinatorial explosion could be addressed by allowing programmers to opt in or out of exhaustivity checks. This almost already exists with `# pyright: ignore[reportMatchNotExhaustive]`, but when `reportUnnecessaryTypeIgnoreComment` is enabled, pyright would still check for exhaustiveness to be able to report whether or not the ignore comment is necessary.
